[hist] Rewrite `TH1::GetCumulative` to produce sensible 2D and 3D cumulative histograms

ShamrockLee opened the pull request:
# This Pull request:

## Changes or fixes:

This PR re-implements the `TH1` method `GetCumulative(forward, suffix)`, which is equivalent to the original implementation (with #11948 applied) on 1D histograms, but produces more sensible result for 2D and 3D histograms. That is,

$$S_{i_x i_y i_z} = \sum_{j_x = n_{x0}}^{i_x} \sum_{j_y = n_{y0}}^{i_y} \sum_{j_z = n_{z0}}^{i_z} a_{j_x j_y j_z}\ (\text{forward})$$

$$S_{i_x i_y i_z} = \sum_{j_x = i_x}^{n_x} \sum_{j_y = i_y}^{n_y} \sum_{j_z = i_z}^{n_z} a_{j_x j_y j_z}\ (\text{backward})$$

To achieve $O(n_x n_y n_z)$ time complexity (instead of $O(n_x^2 n_y^2 n_z^2)$ ), the method is implemented using the [inclusion-exclusion principle](https://en.wikipedia.org/wiki/Inclusion%E2%80%93exclusion_principle) while referencing the content of the previously-computed neighboring bins. Namely,

$$\begin{cases}
S_{i_x i_y i_z} = a_{i_x i_y i_z} + S_{(i_x-1) i_y i_z} + S_{i_x (i_y-1) i_z} + S_{i_x i_y (i_z-1)} - S_{(i_x-1) (i_y-1) i_z} - S_{i_x (i_y-1) (i_z-1)} - S_{(i_x-1) i_y (i_z-1)} + S_{(i_x-1) (i_y-1) (i_z-1)} & (\text{forward}) \\
S_{i_x i_y i_z} = a_{i_x i_y i_z} + S_{(i_x+1) i_y i_z} + S_{i_x (i_y+1) i_z} + S_{i_x i_y (i_z+1)} - S_{(i_x+1) (i_y+1) i_z} - S_{i_x (i_y+1) (i_z+1)} - S_{(i_x+1) i_y (i_z+1)} + S_{(i_x+1) (i_y+1) (i_z+1)} & (\text{backward})
\end{cases}$$

This is useful when computing the selection efficiency of two variables:
For example, one can get a histogram (`*h2_eff_pt_eta`) in which each bin represents the selection efficiency where pt and eta is greater than the lower edges of the bin from the 2D histogram of these variables (`*h2_pt_eta`) with the following code,

```c++
TH2 *h2_eff_pt_eta = h2_pt_eta->GetCumulative(kFALSE, "_efficiency");
h2_eff_pt_eta->Scale(h2_eff_pt_eta->GetBinContent(1, 1));
```

## Checklist:

- [X] tested changes locally
- [X] updated the docs (if necessary)

This PR fixes #

lmoneta commented:
Hi, Thank you @ShamrockLee very much for this contribution!
As a general comment, if this implementation is equivalent, just more efficient than the existing one, then we could just replace it. 
If you think both are useful, we could have two internal (protected) functions for the implementation, but expose them in the same interface eventually adding an option.

ShamrockLee commented:
Just updated according to your suggestion in another PR https://github.com/root-project/root/pull/11948#discussion_r1054468412 .

@lmoneta Thank you for your suggestion and reviewing.

The behavior of the added `TH1::GetCumulativeIncExc` is equivalent to the original `TH1::GetCumulative` *only* for 1D histograms (`fDimension == 1`). The behavioral difference in higher-dimensional histograms is to ensure that the output is still sensible when calculating selection efficiencies.

According to the [documentation](https://root.cern.ch/doc/master/classTH1.html#a5709d315b774a89b1f7e6a71b0e9c380), the original implementation of `GetCumulative` method may not make sense for higher-dimensional use. This would be a good reason to replace the existing method with this one. However, the change could be breaking for people who happen to use the existing method on higher-dimensional histograms.

ShamrockLee commented:
Regarding the efficiency, I don't know what's the proper way to benchmark the cumulative histogram generation. In the following simple benchmark targeting the `hpx` 1D histogram in `tutorials/hsimple.root`, the difference is not much, though `TH1::GetCumulativeIncExc` seems slightly faster.

```
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x372cc00
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00026606600
root [3] .q
[yuehshun@lxplus8s13 root-build]$ cd ../root-build.incexc/
[yuehshun@lxplus8s13 root-build.incexc]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x3c458b0
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH

1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_poroot [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00024825600
root [3] .q
[yuehshun@lxplus8s13 root-build.incexc]$ cd ../root-build
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x2b9a7d0
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_re
solution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00027867100
root [3] .q
[yuehshun@lxplus8s13 root-build]$ cd ../root-build.incexc/
[yuehshun@lxplus8s13 root-build.incexc]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x1f2eaf0
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00024412700
root [3] .q
[yuehshun@lxplus8s13 root-build.incexc]$ cd ../root-build
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x23be190
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
ROOT_prompt_1:1:132: error: no member named 'GetCumulativeIncExc' in 'TH1F'
std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
                                                                                                                              ~~~  ^
root [2] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
[IncrementalJIT] addModule() failed: Duplicate definition of symbol '_ZN21__ROOT_SpecialObjects3hpxE'
root [3] .q
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x1d86820
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00027366700
root [3] .q
[yuehshun@lxplus8s13 root-build]$ cd ../root-build.incexc/
[yuehshun@lxplus8s13 root-build.incexc]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x25389f0
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00022777000
root [3] .q
[yuehshun@lxplus8s13 root-build.incexc]$ cd ../root-build
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x2283130
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00030141800
root [3] .q
[yuehshun@lxplus8s13 root-build]$ cd ../root-build.incexc/
[yuehshun@lxplus8s13 root-build.incexc]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x29b7b20
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::h

igh_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00026997000
root [3] .q
[yuehshun@lxplus8s13 root-build.incexc]$ cd ../root-build
[yuehshun@lxplus8s13 root-build]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x2bec4d0
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulative(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00024567600
root [3] .q
[yuehshun@lxplus8s13 root-build]$ cd ../root-build.incexc/
[yuehshun@lxplus8s13 root-build.incexc]$ ./bin/root -l tutorials/hsimple.root
root [0]
Attaching file tutorials/hsimple.root as _file0...
(TFile *) 0x36d9a20
root [1] std::chrono::time_point<std::chrono::high_resolution_clock> start = std::chrono::high_resolution_clock::now(); TH1 *hpxCumu = hpx->GetCumulativeIncExc(); std::chrono::time_point<std::chrono::high_resolution_clock> stop = std::chrono::high_resolution_clock::now();
root [2] static_cast<std::chrono::duration<double>>(stop - start).count()
(double) 0.00023166700
root [3]
```

ShamrockLee commented:
@lmoneta I just replaced `TH1::GetCumulative` with this new implementation.

Should I document the potentially-breaking change somewhere?

ShamrockLee commented:
Just rebased this PR onto the master branch and resolved the merge conflict.

ShamrockLee committed:
[hist] Rewrite TH1::GetCumulative to produce sensible 2D / 3D histograms

Now the TH1::GetCumulative method is implemented using the
inclusion-exclusion principle, which guarentees that
the content of each bin (S_{i_x,i_y,i_z}) in the cumulative histogram
is the sum of that of all the bins (a_{j_x,j_y,j_z}) whose indices is
no greater than (j_x <= i_x, etc., when forward)
or no less than (j_x >= i_x, etc., when backward)
the indices of the cumulative bin.

Use block comment instead of line comment to avoid trailing backslash
causing line comments to span across lines
(warning: multi-line comment [-Wcomment]).

ShamrockLee commented:
Just did `git commit --amend --reset-author` name from my cybername (Shamrock Lee) to my legal name (Yueh-Shun Li).

ShamrockLee commented:
@lmoneta Are you available to take a look at this one?

ShamrockLee commented:
Changed from line comments to a comment block to fix `warning: multi-line comment [-Wcomment]`.

ShamrockLee commented:
@phsft-bot build

ShamrockLee commented:
@lmoneta @guitargeek, Could you help take another look at this PR?

guitargeek committed:
[hist] Add release note and test for multidimensional TH1::GetCumulative

Add release note and test for multidimensional the TH1::GetCumulative
implemented in the previous commit, and also fix up the doxygen formula.

guitargeek reviewed (approved):
Thank you very much for these developments! I have added a test and release note entries to get this PR to a merge-able state.
